### PR TITLE
feat(esl-utils, esl-traversing-query): update createSequenceFinder, add possibility to find closest element

### LINKED ISSUE #929: [🚀esl-traversing-query]: add ability to find closest element
As an ESL consumer, I want to be able to use the 'closest' traversing operation.

## Operation definition:
`::closet(sel)` = `::is(sel)` or `::parent(sel)`  - find the closest element parent starting and including the current element

## Implementation details

`esl-utils/dom/traversing` should be extended with the `findClosest` function or ability to include the current element for "sequence finder"-based utils.

##Example 
- `TraversingQuery.first($el, '::closest(div)')`
- `findClosest($el, 'div')` or `findPrent($el, 'div', /*Include initial*/ true)`

**Priority**: Normal

### LINKED ISSUE #930: [🚀esl-utils]: add ability to pass predicate to sequence finder
As an ESL consumer, I want to be able to pass the predicate function instead of the selector

**Extended finder API**
```ts
/** Create function that finds next dom element, that matches selector, in the sequence declared by `next` function */
export const createSequenceFinder = (next: (el: Element) => Element | null) => {
  return function (base: Element, sel: string | (el: Element) => boolean): Element | null {
     ...
  };
};
```

**Priority**: Normal

### PR DESCRIPTION
**Pull request by NastaLeo**
[esl-utils]: add ability to pass predicate to createSequenceFinder function
[esl-traversing-query]: add ability to find closest element

In the scope of this PR I reworked some tests related to these topics.

[Closes #930 ](https://github.com/exadel-inc/esl/issues/930)
[Closes #929 ](https://github.com/exadel-inc/esl/issues/929)

### CODE REVIEW

File: src/modules/esl-traversing-query/core/esl-traversing-query.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ type CollectionProcessor = (els: Element[], sel: string) => Element[];
  * - plain CSS selectors
  * - relative selectors (selectors that don't start from a plain selector will use passed base Element as a root)
  * - ::next and ::prev sibling pseudo-selectors
- * - ::parent and ::child pseudo-selectors
+ * - ::parent, closest and ::child pseudo-selectors
```
Comment (dshovchko) [inline src/modules/esl-traversing-query/core/esl-traversing-query.ts:16]:
```suggestion
 * - ::parent, ::closest and ::child pseudo-selectors
```

File: src/modules/esl-traversing-query/core/esl-traversing-query.ts [hunk 2]
```diff
@@ -55,7 +57,7 @@ export class TraversingQuery {
 
   /**
    * @returns RegExp that selects all known processors in query string
-   * e.g. /(::parent|::child|::next|::prev)/
+   * e.g. /(::parent|:closest|::child|::next|::prev)/
```
Comment (dshovchko) [inline src/modules/esl-traversing-query/core/esl-traversing-query.ts:60]:
```suggestion
   * e.g. /(::parent|::closest|::child|::next|::prev)/
```

**Review by dshovchko — APPROVED** [review]

**Review by fshovchko — APPROVED** [review]

File: src/modules/esl-traversing-query/core/esl-traversing-query.ts (hunk 2 above)
Comment (NastaLeo) [reply src/modules/esl-traversing-query/core/esl-traversing-query.ts]: Fixed

**Review by yadamskaya — APPROVED** [review]

**Review by abarmina — APPROVED** [review]

**Review by ala-n — APPROVED** [review]

File: src/modules/esl-utils/dom/traversing.ts [hunk 3]
```diff
@@ -1,13 +1,18 @@
+import type {Predicate} from '../misc/functions';
 /** Check that `nodeA` and `nodeB` are from the same tree path */
 export const isRelativeNode = (nodeA: Node | null, nodeB: Node | null): boolean => {
   return !!(nodeA && nodeB) && (nodeA.contains(nodeB) || nodeB.contains(nodeA));
 };
 
+type iteratorFn = (el: Element) => Element | null;
```
[the remark this answers is not in the document]
Comment (NastaLeo) [reply src/modules/esl-utils/dom/traversing.ts]: Thanks) Fixed

**Pull request merged**